dashboard/app/jobs/evaluate_rubric_job.rb

```diff
@@ -86,6 +86,40 @@ class EvaluateRubricJob < ApplicationJob
     raise exception
   end
 
+  rescue_from(ProfanityFilterException) do |exception|
+    if rack_env?(:development)
+      puts "EvaluateRubricJob Filter Error: #{exception.full_message} Type: #{exception.share_failure.type}"
+    end
+
+    # Record the failure, if we can
+    begin
+      rubric_ai_evaluation = pass_in_or_create_rubric_ai_evaluation(self)
+      rubric_ai_evaluation.status = SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PROFANITY_VIOLATION]
+      rubric_ai_evaluation.save!
+    rescue StandardError
+      # Ignore cascading errors when the rubric record does not exist
+    end
+
+    # We gracefully just fail, here, and we do not file this exception
+  end
+
+  rescue_from(PIIFilterException) do |exception|
+    if rack_env?(:development)
+      puts "EvaluateRubricJob Filter Error: #{exception.full_message} Type: #{exception.share_failure.type}"
+    end
+
+    # Record the failure, if we can
+    begin
+      rubric_ai_evaluation = pass_in_or_create_rubric_ai_evaluation(self)
+      rubric_ai_evaluation.status = SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PII_VIOLATION]
+      rubric_ai_evaluation.save!
+    rescue StandardError
+      # Ignore cascading errors when the rubric record does not exist
+    end
+
+    # We gracefully just fail, here, and we do not file this exception
+  end
```

Comment:
Oh that's something I should confirm. I think I misread some documentation and I did intend to test that. It is intended that one happen and not hit the general case.

Reply:
You can see my confusion: https://edgeapi.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html

> Handlers are inherited. They are searched from right to left, ***from bottom to top***, and up the hierarchy.

I assume they are ALL executed if they match and it isn't just finding the FIRST one. But I'm not sure why it would do the general error first if it goes bottom to top.

Reply:
> You can see my confusion: https://edgeapi.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html
> 
> > Handlers are inherited. They are searched from right to left, _**from bottom to top**_, and up the hierarchy.
> 
> I assume they are ALL executed if they match and it isn't just finding the FIRST one. But I'm not sure why it would do the general error first if it goes bottom to top.

Yes, it is rather confusing and I can't say I totally understand what the docs are saying should happen here. the next line from the docs are:

> The handler of the first class for which exception.is_a?(klass) holds true is the one invoked, if any.

which suggests that only one of your `rescue_from` blocks will get called (in your case the later, more specific ones in case of PIIFilterException). then it says:

> Exceptions raised inside exception handlers are not propagated up.

I am assuming this means that they are not propagated up (within the file) to other rescue_from handlers, since we already know from the existing code that reraised exceptions make it "out" of the rescue_from block and into honeybadger.

Reply:
in any case I'm happy to retract my comment on changing the order based on the possibility of double debug statements, but I do think it is worth confirming that the new handlers are being called and taking effect, and the right things are being logged to honey badger.

Reply:
Ok. Tested this. Learning a lot.

I guarded the setting of the PII_PROFANITY status to happen only if the status is not already FAILURE. If the StandardError happens first, the PII_PROFANITY status does not get set, so the tests fail. For good measure, guarded the normal error path so it doesn't set FAILURE status if it is set to a PII error status.

With `StandardError` on top, tests still pass... **confirming the order is bottom to top**.

Putting the `StandardError` lambda on the /bottom/ causes the two PII failure tests to fail because the exceptions bubble out unexpectedly. That suggests that the `StandardError` happens (bottom-most) and it re-raises and it doesn't perform any of the other `rescue_from`'s.

To sum up:

Defined in this order with print statements:

```
rescue_from(ProfanityFilterException)
rescue_from(PIIFilterException)
rescue_from(StandardError)
```

Produced:

```
StandardError
StandardError
StandardError
```

So they match bottom to top and once they match, they do not continue to match. So you want your most specific cases on the bottom and catch-alls at the top of the module.

Reply:
https://www.vzhima.com/2019/12/02/rails-multiple-rescue-from.html

Reply:
Wonderful! Thank you for getting to the bottom of this! 

> Produced:
> 
> ```
> StandardError
> StandardError
> StandardError
> ```

What does each `StandardError` represent here?